sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,530 @@
+        }
+    }
+
+    private byte[] getWrappedKey() {
+        try (IndexInput in = super.openInput("keyfile", new IOContext())) {
+            int size = in.readInt();
+            byte[] ret = new byte[size];
+            in.readBytes(ret, 0, ret.length);
+            return ret;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param source the file to be renamed
+     * @param dest the new file name
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        super.rename(source, dest);
+        if (!(source.contains("segments_") || source.endsWith(".si"))) ivMap.put(
+            getDirectory() + "/" + dest,
+            ivMap.remove(getDirectory() + "/" + source)
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for reading
+     * @param context the IO context
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.openInput(name, context);
+        ensureOpen();
+        ensureCanRead(name);
+        Path path = getDirectory().resolve(name);
+        FileChannel fc = FileChannel.open(path, StandardOpenOption.READ);
+        boolean success = false;
+        try {
+            Cipher cipher = CipherFactory.getCipher(provider);
+            String ivEntry = ivMap.get(getDirectory() + "/" + name);
+            if (ivEntry == null) throw new IOException("failed to open file. " + name);
+            byte[] iv = Base64.getDecoder().decode(ivEntry);
+            CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.DECRYPT_MODE, 0);
+            final IndexInput indexInput;
+            indexInput = new CryptoBufferedIndexInput("CryptoBufferedIndexInput(path=\"" + path + "\")", fc, context, cipher, this);
+            success = true;
+            return indexInput;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(fc);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for writing
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.createOutput(name, context);
+        ensureOpen();
+        OutputStream fos = Files.newOutputStream(directory.resolve(name), StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+        Cipher cipher = CipherFactory.getCipher(provider);
+        SecureRandom random = Randomness.createSecure();
+        byte[] iv = new byte[CipherFactory.IV_ARRAY_LENGTH];
+        random.nextBytes(iv);
+        if (dataKey == null) throw new RuntimeException("dataKey is null!");
```

Comment:
I suggest creating a new extended exception type unique to the plugin. You can pick the name etc. but just something so you know the plugin the exception comes from. 

For the error message: 1) we should probably use constants files for storing strings like this as it makes testing a bit easier and helps with any translators run over the files; 2) could you swap the exception message to share info on the result for ex: "Crypto plugin failed to createOutput for reason: " + dataKey is null.